controller/object_store_controller.go

```diff
@@ -0,0 +1,1159 @@
+	}
+	return nil
+}
+
+func (osc *ObjectStoreController) getOrCreateVolume(
+	store *longhorn.ObjectStore,
+	pvc *corev1.PersistentVolumeClaim,
+) (*longhorn.Volume, *longhorn.ObjectStore, error) {
+	vol, err := osc.ds.GetVolume(genPVName(store))
+	if err == nil {
+		return vol, store, nil
+	} else if datastore.ErrorIsNotFound(err) {
+		vol, err = osc.createVolume(store, pvc)
+		if err != nil {
+			return nil, store, errors.Wrap(err, "failed to create longhorn volume")
+		} else if store.Status.State != longhorn.ObjectStoreStateStarting {
+			store.Status.State = longhorn.ObjectStoreStateStarting
+		}
+		return vol, store, nil
+	}
+	return nil, store, err
+}
+
+func (osc *ObjectStoreController) checkVolume(vol *longhorn.Volume) error {
+	if vol.Status.Robustness == longhorn.VolumeRobustnessFaulted {
+		return errors.New(fmt.Sprintf("volume %v has failed", vol.Name))
+	}
+	return nil
+}
+
+func (osc *ObjectStoreController) getOrCreatePV(
+	store *longhorn.ObjectStore,
+	volume *longhorn.Volume,
+) (*corev1.PersistentVolume, *longhorn.ObjectStore, error) {
+	pv, err := osc.ds.GetPersistentVolume(genPVName(store))
+	if err == nil {
+		return pv, store, nil
+	} else if datastore.ErrorIsNotFound(err) {
+		pv, err = osc.createPV(store, volume)
+		if err != nil {
+			return nil, store, errors.Wrap(err, "failed to create persistent volume")
+		} else if store.Status.State != longhorn.ObjectStoreStateStarting {
+			store.Status.State = longhorn.ObjectStoreStateStarting
+		}
+		return pv, store, nil
+	}
+	return nil, store, err
+}
+
+func (osc *ObjectStoreController) checkPV(pv *corev1.PersistentVolume) error {
+	if pv.Status.Phase != corev1.VolumeBound {
+		return errors.New(fmt.Sprintf("PV %v not bound", pv.Name))
+	}
+	return nil
+}
+
+func (osc *ObjectStoreController) getOrCreateDeployment(store *longhorn.ObjectStore) (*appsv1.Deployment, *longhorn.ObjectStore, error) {
+	dpl, err := osc.ds.GetDeployment(store.Name)
+	if err == nil {
+		return dpl, store, nil
+	} else if datastore.ErrorIsNotFound(err) {
+		dpl, err = osc.createDeployment(store)
+		if err != nil {
+			return nil, store, errors.Wrap(err, "failed to create deployment")
+		} else if store.Status.State != longhorn.ObjectStoreStateStarting {
+			store.Status.State = longhorn.ObjectStoreStateStarting
+		}
+		return dpl, store, nil
+	}
+	return nil, store, err
+}
+
+func (osc *ObjectStoreController) checkDeployment(deployment *appsv1.Deployment, store *longhorn.ObjectStore) error {
+	if *deployment.Spec.Replicas != 1 {
+		deployment.Spec.Replicas = int32Ptr(1)
+		osc.ds.UpdateDeployment(deployment)
+		return errors.New("deployment just scaled")
+	} else if deployment.Status.Replicas == 0 || deployment.Status.UnavailableReplicas > 0 {
+		return errors.New("deployment not ready")
+	}
+
+	if store.Spec.Image != "" && deployment.Spec.Template.Spec.Containers[0].Image != store.Spec.Image {
+		deployment.Spec.Template.Spec.Containers[0].Image = store.Spec.Image
```

Comment:
This might be problematic as we cannot always guarantee that containers[0] is the s3gw-backend and containers[1] is the s3gw-ui. Some clusters even automatically injected some additional containers into every pod. Suggesting search for the container name `s3gw` and `s3gw-ui` before assigning the image